src/ucp/core/ucp_am.c

```diff
@@ -90,29 +90,51 @@ UCS_PROFILE_FUNC_VOID(ucp_am_data_release, (worker, data),
     UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(worker);
 }
 
+static void ucp_worker_am_init_handler(ucp_worker_h worker, uint16_t id,
+                                       void *context, unsigned flags,
+                                       ucp_am_callback_t cb_old,
+                                       ucp_am_recv_callback_t cb)
+{
+    ucp_am_entry_t *am_cb = &ucs_array_elem(&worker->am, id);
+
+    am_cb->context = context;
+    am_cb->flags   = flags;
+
+    if (cb_old != NULL) {
+        ucs_assert(cb == NULL);
+        am_cb->cb_old = cb_old;
+    } else {
+        ucs_assert(cb_old == NULL);
+        am_cb->cb     = cb;
+    }
+}
+
 static ucs_status_t ucp_worker_set_am_handler_common(ucp_worker_h worker,
-                                                     uint16_t id)
+                                                     uint16_t id,
+                                                     unsigned flags)
 {
-    size_t num_entries;
     ucs_status_t status;
     unsigned i, capacity;
 
     UCP_CONTEXT_CHECK_FEATURE_FLAGS(worker->context, UCP_FEATURE_AM,
                                     return UCS_ERR_INVALID_PARAM);
 
+    if (flags & UCP_AM_CB_PRIV_FLAG_NBX) {
```

Comment:
maybe add something like UCP_AM_CB_FIRST_PRIV_FLAG and check that flags < UCP_AM_CB_FIRST_PRIV_FLAG ?